File: Windows/Resources/SDEngine/WeatherInformation.cs
using SDEngine.Convertions;
using SDEngine.Convertions.Enumerations;
using SDEngine.Memory;
using System;
using System.Diagnostics;
using System.Xml.Linq;

namespace SDEngine {
    public class WeatherInformation {
        private XDocument sourceDocument;

        public double Latitude {
            get; internal set;
        }

        public double Longitude {
            get; internal set;
        }

        public double Elevation {
            get; internal set;
        }

        public string ConditionString {
            get; internal set;
        }

        public double Temperature {
            get; internal set;
        }

        public double FeelsLike {
            get; internal set;
        }

        public string Humidity {
            get; internal set;
        }

        public string WindDirection {
            get; internal set;
        }

        public double WindSpeed {
            get; internal set;
        }

        public double WindGustSpeed {
            get; internal set;
        }

        public double Pressure {
            get; internal set;
        }

        public Uri ForecastUrl {
            get; internal set;
        }

        public Uri HistoryUrl {
            get; internal set;
        }

        public void Update() {
            XElement root = sourceDocument.Element("response"),
                     observation = root.Element("current_observation"),
                     location = observation.Element("display_location");
            Temperature = (Manager.TempUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "temp_f").Value),
                    2) :
                (Manager.TempUnit == 1) ?
                    Math.Round(
                        double.Parse(
                            observation.Element(
                                "temp_c").Value),
                        2) :
                    Math.Round(
                        Converter.Convert(
                            TemperatureUnit.Kelvin,
                            TemperatureUnit.Celsius,
                            double.Parse(
                                observation.Element(
                                    "temp_c").Value)),
                        2);
            FeelsLike = (Manager.TempUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "feelslike_f").Value),
                    2) :
                (Manager.TempUnit == 1) ?
                    Math.Round(
                        double.Parse(
                            observation.Element(
                                "feelslike_c").Value),
                        2) :
                    Math.Round(
                        Converter.Convert(
                            TemperatureUnit.Celsius,
                            TemperatureUnit.Kelvin,
                            double.Parse(
                                observation.Element(
                                    "feelslike_c").Value)),
                        2);
            WindSpeed = (Manager.SpeedUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_mph").Value),
                    2) :
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_kph").Value),
                    2);
            WindGustSpeed = (Manager.SpeedUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_gust_mph").Value),
                    2) :
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_gust_kph").Value),
                    2);
            Pressure = (Manager.PressureUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "pressure_in").Value),
                    2) :
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "pressure_mb").Value),
                    2);
            Manager.Temp = Temperature;
        }

        public WeatherInformation(string xml) {
            sourceDocument = XDocument.Parse(xml);
            XElement root = sourceDocument.Element("response"),
                     observation = root.Element("current_observation"),
                     location = observation.Element("display_location");
            Latitude = Math.Round(
                double.Parse(
                    location.Element(
                        "latitude").Value),
                2);
            Longitude = Math.Round(
                double.Parse(
                    location.Element(
                        "longitude").Value),
                2);
            Elevation = Math.Round(
                double.Parse(
                    location.Element(
                        "elevation").Value),
                2);
            ConditionString = observation.Element("weather").Value;
            Debug.WriteLine("Temperature");
            Temperature = (Manager.TempUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "temp_f").Value),
                    2) :
                (Manager.TempUnit == 1) ?
                    Math.Round(
                        double.Parse(
                            observation.Element(
                                "temp_c").Value),
                        2) :
                    Math.Round(
                        Converter.Convert(
                            TemperatureUnit.Kelvin,
                            TemperatureUnit.Celsius,
                            double.Parse(
                                observation.Element(
                                    "temp_c").Value)),
                        2);
            Debug.WriteLine("FeelsLike");
            FeelsLike = (Manager.TempUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "feelslike_f").Value),
                    2) :
                (Manager.TempUnit == 1) ?
                    Math.Round(
                        double.Parse(
                            observation.Element(
                                "feelslike_c").Value),
                        2) :
                    Math.Round(
                        Converter.Convert(
                            TemperatureUnit.Celsius,
                            TemperatureUnit.Kelvin,
                            double.Parse(
                                observation.Element(
                                    "feelslike_c").Value)),
                        2);
            Humidity = observation.Element("relative_humidity").Value;
            WindDirection = observation.Element("wind_dir").Value;
            Debug.WriteLine("WindSpeed");
            WindSpeed = (Manager.SpeedUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_mph").Value),
                    2) :
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_kph").Value),
                    2);
            Debug.WriteLine("WindGustSpeed");
            WindGustSpeed = (Manager.SpeedUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_gust_mph").Value),
                    2) :
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "wind_gust_kph").Value),
                    2);
            Debug.WriteLine("Pressure");
            Pressure = (Manager.PressureUnit == 0) ?
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "pressure_in").Value),
                    2) :
                Math.Round(
                    double.Parse(
                        observation.Element(
                            "pressure_mb").Value),
                    2);
            ForecastUrl = new Uri(observation.Element("forecast_url").Value, UriKind.Absolute);
            HistoryUrl = new Uri(observation.Element("history_url").Value, UriKind.Absolute);
            Manager.Temp = Temperature;
            Manager.csource = xml;
        }
    }
}

File: Android/Resources/SDEngine/src/sdengine/WeatherInformation.java
package sdengine;

import org.json.JSONException;
import org.json.JSONObject;

import sdengine.convertions.Converter;
import sdengine.convertions.enumerations.TemperatureUnit;
import android.net.Uri;
import android.util.Log;

public class WeatherInformation {
	private String jsonSource;
	
	public Double latitude;
	public Double longitude;
	public Double elevation;
	public String condtionString;
	public Double temperature;
	public Double feelsLike;
	public String humidity;
	public String windDirection;
	public Double windSpeed;
	public Double windGustSpeed;
	public String pressure;
	public Uri forecastUri;
	public Uri historicUri;
	
	public void update(int tempUnit, int speedUnit, int psrUnit) {
	    try {
	    	JSONObject data = new JSONObject(jsonSource);
			JSONObject obs = data.getJSONObject("current_observation");
			temperature = (tempUnit == 0) ? Converter.round(obs.getDouble("temp_f")) : (tempUnit == 1) ? Converter.round(obs.getDouble("temp_c")) : Converter.round(Converter.convert(TemperatureUnit.Celsius, TemperatureUnit.Kelvin, obs.getDouble("temp_c")));
			feelsLike = (tempUnit == 0) ? Converter.round(obs.getDouble("feelslike_f")) : (tempUnit == 1) ? Converter.round(obs.getDouble("feelslike_c")) : Converter.round(Converter.convert(TemperatureUnit.Celsius, TemperatureUnit.Kelvin, obs.getDouble("feelslike_c")));
			windSpeed = (speedUnit == 0) ? Converter.round(obs.getDouble("wind_mph")) : Converter.round(obs.getDouble("wind_kph"));
			windGustSpeed = (speedUnit == 0) ? Converter.round(obs.getDouble("wind_gust_mph")) : Converter.round(obs.getDouble("wind_gust_kph"));
			pressure = (psrUnit == 0) ? obs.getString("pressure_in") : obs.getString("pressure_mb");
	    } catch (JSONException e) {
	    	e.printStackTrace();
	    }
	}
	
	public WeatherInformation(String json, int tempUnit, int speedUnit, int psrUnit) {
		try {
			Log.d("sdengine", json);
			JSONObject data = new JSONObject(json);
			JSONObject obs = data.getJSONObject("current_observation");
			JSONObject loc = obs.getJSONObject("display_location");
			latitude = Converter.round(Double.parseDouble(loc.getString("latitude")));
			longitude = Converter.round(Double.parseDouble(loc.getString("longitude")));
			elevation = Converter.round(Double.parseDouble(loc.getString("elevation")));
			condtionString = (obs.getString("weather") == "") ? "Unavailable" : obs.getString("weather");
			temperature = (tempUnit == 0) ? Converter.round(obs.getDouble("temp_f")) : (tempUnit == 1) ? Converter.round(obs.getDouble("temp_c")) : Converter.round(Converter.convert(TemperatureUnit.Celsius, TemperatureUnit.Kelvin, obs.getDouble("temp_c")));
			feelsLike = (tempUnit == 0) ? Converter.round(obs.getDouble("feelslike_f")) : (tempUnit == 1) ? Converter.round(obs.getDouble("feelslike_c")) : Converter.round(Converter.convert(TemperatureUnit.Celsius, TemperatureUnit.Kelvin, obs.getDouble("feelslike_c")));
			humidity = obs.getString("relative_humidity");
			windSpeed = (speedUnit == 0) ? Converter.round(obs.getDouble("wind_mph")) : Converter.round(obs.getDouble("wind_kph"));
			windGustSpeed = (speedUnit == 0) ? Converter.round(obs.getDouble("wind_gust_mph")) : Converter.round(obs.getDouble("wind_gust_kph"));
			pressure = (psrUnit == 0) ? obs.getString("pressure_in") : obs.getString("pressure_mb");
			forecastUri = Uri.parse(obs.getString("forecast_url"));
			historicUri = Uri.parse(obs.getString("history_url"));
			jsonSource = json;
		}
		catch (JSONException e) {
			e.printStackTrace();
		}
	}
}
